src/OpenTelemetry/Trace/Export/BatchingActivityProcessor.cs

```diff
@@ -0,0 +1,257 @@
+            this.maxQueueSize = maxQueueSize;
+            this.scheduledDelay = scheduledDelay;
+            this.exporterTimeout = exporterTimeout;
+            this.maxExportBatchSize = maxExportBatchSize;
+
+            this.cts = new CancellationTokenSource();
+            this.exportQueue = new ConcurrentQueue<Activity>();
+
+            // worker task that will last for lifetime of processor.
+            // Threads are also useless as exporter tasks run in thread pool threads.
+            Task.Run(() => this.Worker(this.cts.Token), this.cts.Token);
+        }
+
+        /// <inheritdoc/>
+        public override void OnStart(Activity activity)
+        {
+        }
+
+        /// <inheritdoc/>
+        public override void OnEnd(Activity activity)
+        {
+            if (this.stopping)
+            {
+                return;
+            }
+
+            // because of race-condition between checking the size and enqueueing,
+            // we might end up with a bit more activities than maxQueueSize.
+            // Let's just tolerate it to avoid extra synchronization.
+            if (this.currentQueueSize >= this.maxQueueSize)
+            {
+                OpenTelemetrySdkEventSource.Log.SpanProcessorQueueIsExhausted();
+                return;
+            }
+
+            Interlocked.Increment(ref this.currentQueueSize);
+
+            this.exportQueue.Enqueue(activity);
+        }
+
+        /// <inheritdoc/>
+        public override async Task ShutdownAsync(CancellationToken cancellationToken)
+        {
+            if (!this.stopping)
+            {
+                this.stopping = true;
+
+                // This will stop the loop after current batch finishes.
+                this.cts.Cancel(false);
+                this.cts.Dispose();
+                this.cts = null;
+
+                // if there are more items, continue until cancellation token allows
+                while (this.currentQueueSize > 0 && !cancellationToken.IsCancellationRequested)
+                {
+                    await this.ExportBatchAsync(cancellationToken).ConfigureAwait(false);
+                }
+
+                await this.exporter.ShutdownAsync(cancellationToken);
+
+                // there is no point in waiting for a worker task if cancellation happens
+                // it's dead already or will die on the next iteration on its own
+
+                // ExportBatchAsync must never throw, we are here either because it was cancelled
+                // or because there are no items left
+                OpenTelemetrySdkEventSource.Log.ShutdownEvent(this.currentQueueSize);
+            }
+        }
+
+        public void Dispose()
+        {
+            this.Dispose(true);
+        }
+
+        protected virtual void Dispose(bool isDisposing)
+        {
+            if (!this.stopping)
+            {
+                this.ShutdownAsync(CancellationToken.None).ContinueWith(_ => { }).GetAwaiter().GetResult();
```

Comment:
This could introduce exception which is against the [.NET guideline](docs.microsoft.com/en-us/dotnet/api/system.idisposable.dispose): _The object must not throw an exception if its Dispose method is called multiple times_.

Reason: `ShutdownAsync` is not thread safe, race condition could happen for `this.stopping`, which would re-enter the following code with `this.cts == null`:
```csharp
this.cts.Cancel(false);
this.cts.Dispose();
this.cts = null;
```

Reply:
What's with `ContinueWith(_ => { })` is that to suppress any exception?

Reply:
Yes. This will suppress an exception.

Reply:
@reyang @CodeBlanch @cijothomas 
We have similar dispose pattern in SimpleSpanProcessor, SimpleActivityProcessor, BatchingSpanProcessor. Should we create a new issue to address it? or we consider fixing as a part of this PR?

Reply:
i'd open a new issue and merge this one. So that we can make progress and keep track of fixing the underlying issue.

Reply:
Created an issue https://github.com/open-telemetry/opentelemetry-dotnet/issues/769 to track it.